include/pmacc/meta/ExecuteIf.hpp

```diff
@@ -0,0 +1,120 @@
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <utility>
+
+namespace pmacc
+{
+    namespace meta
+    {
+        namespace detail
+        {
+            /** Conditional functor execution
+             *
+             * @tparam T_condition condition for executing a functor
+             */
+            template<bool T_condition>
+            struct IfElse;
+
+            /** Specialization if condition is true
+             *
+             * Execute trueFunctor passed to operator()
+             */
+            template<>
+            struct IfElse<true>
+            {
+                template<typename T_ConditionTrueFunctor, typename T_ConditionFalseFunctor, typename... T_Args>
+                void operator()(T_ConditionTrueFunctor&& trueFunctor, T_ConditionFalseFunctor&&, T_Args&&... args)
+                    const
+                {
+                    trueFunctor(std::forward<T_Args>(args)...);
+                }
+            };
+
+            /** Specialization if condition is false
+             *
+             * Execute falseFunctor passed to operator()
+             */
+            template<>
+            struct IfElse<false>
+            {
+                template<typename T_ConditionTrueFunctor, typename T_ConditionFalseFunctor, typename... T_Args>
+                void operator()(T_ConditionTrueFunctor&&, T_ConditionFalseFunctor&& falseFunctor, T_Args&&... args)
+                    const
+                {
+                    falseFunctor(std::forward<T_Args>(args)...);
+                }
+            };
+        } // namespace detail
+
+        /** Conditional functor execution
+         *
+         * Evaluate and execute a functor based on the conditional compile-time argument.
+         * The functor implementations must have valid C++ syntax but is only evaluated if executed.
+         *
+         * @tparam T_condition execute trueFunctor if the condition is true, else no operation is performed
+         * @tparam T_ConditionTrueFunctor functor type, the functor is only evaluated if T_condition is true
+         * @tparam T_Args types of the arguments for trueFunctor call
+         * @param trueFunctor functor instance which is called with the arguments args if T_condition is true
+         * @param args arguments forwarded to the executed functor
+         */
+        template<bool T_condition, typename T_ConditionTrueFunctor, typename... T_Args>
+        constexpr void executeIf(T_ConditionTrueFunctor&& trueFunctor, T_Args&&... args)
```

Comment:
This works in CUDA and HIP as the function is `constexpr`. But probably it is more generaly to add `HDINLINE` here and for the `IfElse` function

Reply:
Unrelated: i think `invokeIf` would be a  more suitable name as here we are running some function which is commonly named _invoking_, see [std::invoke](https://en.cppreference.com/w/cpp/utility/functional/invoke)